### PR TITLE
FP S1128: ignore unused imported symbols denoting JSX factories and JSX fragment factories

### LINKED ISSUE #2701: FP S1128: ignore unused imported symbols denoting JSX factories and JSX fragment factories
JSX requires `h` function to be imported, however we report the import to be unused. 

We should detect if `jsxFactory` setting is present in `tsconfig` and ignore `h` function import.

See more details in the user report

https://community.sonarsource.com/t/false-positive-remove-this-unused-import-of-h-in-all-stenciljs-components/42367

### PR DESCRIPTION
**Pull request by yassin-kammoun-sonarsource**
Fixes #2701

### CODE REVIEW

File: eslint-bridge/tests/rules/unused-import.test.ts [hunk 1]
```diff
@@ -258,3 +259,57 @@ ruleTesterTS.run('Unnecessary imports should be removed', rule, {
...
+});
+
+const filename = path.resolve(`${__dirname}/../fixtures/unused-import/sample.lint.tsx`);
+
+ruleTesterJsxFactory.run('Unused imports denoting jsx factory should be ignored', rule, {
+  valid: [
+    {
+      filename,
+      code: `
+        import { h } from 'jsxlib';
```
Comment (vilchik-elena) [inline eslint-bridge/tests/rules/unused-import.test.ts:280]: it does not matter from which lib it's imported, right? (in the docs it's `preact`)

Comment (vilchik-elena) [inline eslint-bridge/tests/rules/unused-import.test.ts:280]: let's add test to show that we report on `h` when jsxFactory is not defined (or we can use simply js project)

File: eslint-bridge/tests/rules/unused-import.test.ts [hunk 2]
```diff
@@ -258,3 +259,57 @@ ruleTesterTS.run('Unnecessary imports should be removed', rule, {
...
+        export class Component {
+          render() {
+            return <div>Hello, world!</div>
```
Comment (vilchik-elena) [inline eslint-bridge/tests/rules/unused-import.test.ts:283]: I am not sure about JSX in the code here, on one side we imitate real code but then it's not clear from the test that we don't care about jsx presence. How about having both?)

File: eslint-bridge/src/rules/unused-import.ts [hunk 3]
```diff
@@ -48,6 +49,15 @@ export const rule: Rule.RuleModule = {
       return variable.name === 'jsx' && isJsxPragmaSet;
     }
 
+    function getJsxFactory() {
+      const parserServices = context.parserServices;
+      if (isRequiredParserServices(parserServices)) {
+        const compilerOptions = parserServices.program.getCompilerOptions();
+        return compilerOptions.jsxFactory;
```
Comment (vilchik-elena) [inline eslint-bridge/src/rules/unused-import.ts:56]: let's also support this property https://www.typescriptlang.org/tsconfig#jsxFragmentFactory (ticket and PR title to be updated then)

**Review by vilchik-elena — CHANGES_REQUESTED** [review]

File: eslint-bridge/tests/rules/unused-import.test.ts (hunk 1 above)
Comment (yassin-kammoun-sonarsource) [reply eslint-bridge/tests/rules/unused-import.test.ts:280]: Indeed, it doesn't matter. I will rename the lib to avoid the confusion.

**Review by vilchik-elena — APPROVED** [review]

**Pull request merged**